
<!doctype html>
<html lang="en" class="no-js">
  <head>
    

      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://sib-swiss.github.io/NGS-variants-training/2024.9/day2/variant_calling/">
      
      
        <link rel="prev" href="../variant_calling_preparation/">
      
      
        <link rel="next" href="../visualisation/">
      
      
      <link rel="icon" href="../../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.33">
    
<script type="application/ld+json">
  {
    "@context": "https://schema.org/",
    "@type": "LearningResource",
    "@id": "https://sib-swiss.github.io/NGS-variants-training",
    "http://purl.org/dc/terms/conformsTo": {
        "@type": "CreativeWork",
        "@id": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE"
    },
    "description": "Course on detecting variants from next generation sequencing data",
    "keywords": "NGS, variant, SNP, INDEL, sequencing, SNV, gatk",
    "name": "NGS - variant analysis",
    "about": [
        {
            "@type": "DefinedTerm",
            "@id": "http://edamontology.org/topic_3168",
            "inDefinedTermSet": "http://edamontology.org",
            "termCode": "topic_3168",
            "name": "Sequencing"
        },
        {
            "@type": "DefinedTerm",
            "@id": "http://edamontology.org/topic_3227",
            "inDefinedTermSet": "http://edamontology.org",
            "termCode": "topic_3227",
            "name": "Variant calling"
        }
    ],
    "abstract": "The detection of genetic variation is of major interest in various disciplines spanning from ecology and evolution research to inherited disease discovery and precision oncology. Next generation sequencing (NGS) methods are very powerful for the detection of genomic variants. Thanks to its throughput and cost-efficiency it enables the detection of a large number of variants in a large number of samples. In this two-day course we will cover the steps from read alignment to variant calling and annotation. We will mainly focus on the detection of germline mutations by following the GATK best practices. ",
    "audience": "Biologists working with NGS data",
    "author": [
        {
            "@type": "Person",
            "name": "Geert van Geest",
            "email": "geert.vangeest@sib.swiss"
        },
        {
            "@type": "Person",
            "name": "Patricia Palagi"
        },
        {
            "@type": "Organization",
            "name": "SIB Swiss Institute of Bioinformatics"
        }
    ],
    "identifier": "https://doi.org/10.5281/zenodo.6457818",
    "license": "https://creativecommons.org/licenses/by/4.0/",
    "teaches": [
        "Understand important aspects of NGS and read alignment for variant analysis",
        "Perform a read alignment ready for variant analysis",
        "Perform variant calling according to GATK best practices",
        "Perform a variant annotation"
    ]
}
</script>

    
      
        <title>Variant calling - NGS - variant analysis</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.3cba04c6.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
<!-- Matomo -->
<script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
      var u="https://matomo.sib.swiss/";
      _paq.push(['setTrackerUrl', u+'matomo.php']);
      _paq.push(['setSiteId', '220']);
      var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
      g.async=true; g.src=u+'matomo.js'; s.parentNode.insertBefore(g,s);
    })();
  </script>
  <!-- End Matomo Code -->
  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#learning-outcomes" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
        <aside class="md-banner">
          <div class="md-banner__inner md-grid md-typeset">
            
              <button class="md-banner__button md-icon" aria-label="Don't show this again">
                
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
              </button>
            
            
  <b>Issues, questions or feedback? &#128591; &#128640; </b> Use the "Questions and Answers" section at the bottom &#128071; of every page

          </div>
          
            <script>var content,el=document.querySelector("[data-md-component=announce]");el&&(content=el.querySelector(".md-typeset"),__md_hash(content.innerHTML)===__md_get("__announce")&&(el.hidden=!0))</script>
          
        </aside>
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
          <aside class="md-banner md-banner--warning">
            <div class="md-banner__inner md-grid md-typeset">
              
  You're not viewing the latest version.
  <a href="../../.."> 
    <strong>Click here to go to latest.</strong>
  </a>

            </div>
            <script>var el=document.querySelector("[data-md-component=outdated]"),outdated=__md_get("__outdated",sessionStorage);!0===outdated&&el&&(el.hidden=!1)</script>
          </aside>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="NGS - variant analysis" class="md-header__button md-logo" aria-label="NGS - variant analysis" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            NGS - variant analysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Variant calling
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/sib-swiss/NGS-variants-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/NGS-variants-training
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="NGS - variant analysis" class="md-nav__button md-logo" aria-label="NGS - variant analysis" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    NGS - variant analysis
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/sib-swiss/NGS-variants-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/NGS-variants-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../precourse/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Precourse preparations
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../course_schedule/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Course schedule
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Day 1
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Day 1
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/server_login/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Setup
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/reproducibility/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Reproducibility
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/alignment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Read alignment - basics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/alignment_advanced/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Read alignment - advanced
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Day 2
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Day 2
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variant_calling_preparation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant calling - preparation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Variant calling
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Variant calling
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    <span class="md-ellipsis">
      Learning outcomes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    <span class="md-ellipsis">
      Material
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    <span class="md-ellipsis">
      Exercises
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-variant-calling" class="md-nav__link">
    <span class="md-ellipsis">
      1. Variant calling
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Variant calling">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#calculating-pl-and-gq-by-hand" class="md-nav__link">
    <span class="md-ellipsis">
      Calculating PL and GQ by hand
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#calling-variants-with-gatk" class="md-nav__link">
    <span class="md-ellipsis">
      Calling variants with GATK
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-combining-gvcfs" class="md-nav__link">
    <span class="md-ellipsis">
      2. Combining GVCFs
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../visualisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Visualisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../filtering_evaluation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Filtering & evaluation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Annotation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    <span class="md-ellipsis">
      Learning outcomes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    <span class="md-ellipsis">
      Material
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    <span class="md-ellipsis">
      Exercises
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-variant-calling" class="md-nav__link">
    <span class="md-ellipsis">
      1. Variant calling
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Variant calling">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#calculating-pl-and-gq-by-hand" class="md-nav__link">
    <span class="md-ellipsis">
      Calculating PL and GQ by hand
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#calling-variants-with-gatk" class="md-nav__link">
    <span class="md-ellipsis">
      Calling variants with GATK
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-combining-gvcfs" class="md-nav__link">
    <span class="md-ellipsis">
      2. Combining GVCFs
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
  
                  

  
  


  <h1>Variant calling</h1>

<h2 id="learning-outcomes">Learning outcomes</h2>
<p><strong>After having completed this chapter you will be able to:</strong></p>
<ul>
<li>Perform basic calculations regarding the genotype likelihood of individual variants</li>
<li>Follow <code>gatk</code> best practices workflow to perform a variant analysis by:<ul>
<li>Calling variants with <code>gatk HaplotypeCaller</code></li>
<li>Combining multiple <code>vcf</code> files into a single <code>vcf</code> file</li>
</ul>
</li>
<li>Perform basic operations to get statistics of a <code>vcf</code> file</li>
</ul>
<h2 id="material">Material</h2>
<p>The <a href="https://www.biorxiv.org/content/10.1101/201178v3">paper</a> on genomic variant call format (gVCF)</p>
<p><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035535932-Germline-short-variant-discovery-SNPs-Indels-">GATK best practices germline short variant workflow</a>:</p>
<figure>
  <img src="../../assets/images/gatk_germline.png" width="800"/>
</figure>

<h2 id="exercises">Exercises</h2>
<h3 id="1-variant-calling">1. Variant calling</h3>
<h4 id="calculating-pl-and-gq-by-hand">Calculating PL and GQ by hand</h4>
<p>Here&rsquo;s a function in R to calculate genotype likelihoods as described in Li H.  Bioinformatics. 2011;27:2987–93 (assuming equal base error probabilities for all reads):</p>
<div class="highlight"><pre><span></span><code><span class="n">genotype_likelihood</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">m</span><span class="p">,</span><span class="n">g</span><span class="p">,</span><span class="n">e</span><span class="p">,</span><span class="n">ref</span><span class="p">,</span><span class="n">alt</span><span class="p">){</span>
<span class="w">  </span><span class="p">(((</span><span class="n">m</span><span class="o">-</span><span class="n">g</span><span class="p">)</span><span class="o">*</span><span class="n">e</span><span class="o">+</span><span class="n">g</span><span class="o">*</span><span class="p">(</span><span class="m">1</span><span class="o">-</span><span class="n">e</span><span class="p">))</span><span class="o">^</span><span class="n">alt</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="p">((</span><span class="n">m</span><span class="o">-</span><span class="n">g</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="m">1</span><span class="o">-</span><span class="n">e</span><span class="p">)</span><span class="o">+</span><span class="n">g</span><span class="o">*</span><span class="n">e</span><span class="p">)</span><span class="o">^</span><span class="n">ref</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">m</span><span class="o">^</span><span class="p">(</span><span class="n">ref</span><span class="o">+</span><span class="n">alt</span><span class="p">))</span>
<span class="p">}</span>
</code></pre></div>
<p>Where:</p>
<ul>
<li><code>m</code> : ploidy</li>
<li><code>g</code> : number of alternative alleles</li>
<li><code>e</code> : base error probability</li>
<li><code>ref</code> : number of reference alleles counted</li>
<li><code>alt</code> : number of alternative alleles counted</li>
</ul>
<p><strong>Exercise:</strong> In the <code>scripts</code> directory, create a script called <code>calculate_genotype_likelihoods.R</code>. Copy-paste the above function to the script, and use it to calculate the three genotype likelihoods (for g = 0, g = 1 and g = 2) for a case where we count 22 reference alleles and 4 alternative alleles (so a coverage of 26), and base error probability of 0.01. Calculate the PL values (<code>-10*log10(likelihood)</code>) for each genotype. </p>
<div class="admonition tip">
<p class="admonition-title">Using VScode with R</p>
<p>In order to easily interact with your R script, you can do the following:</p>
<ul>
<li>Open the R script in VS code</li>
<li>In the terminal, type <code>R</code> to start the R console</li>
<li>Select the code you&rsquo;d like to run in the R script</li>
<li>Type <span class="keys"><kbd class="key-control">Ctrl</kbd><span>+</span><kbd class="key-enter">Enter</kbd></span> to send it to the console</li>
<li>After you have finished, type <code>quit()</code> in the R console. </li>
</ul>
</div>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><pre><span></span><code><span class="c1"># For g = 0 (i.e. 0 alternative alleles)</span>
<span class="m">-10</span><span class="o">*</span><span class="nf">log10</span><span class="p">(</span><span class="nf">genotype_likelihood</span><span class="p">(</span><span class="n">m</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">g</span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">e</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.01</span><span class="p">,</span><span class="w"> </span><span class="n">ref</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">22</span><span class="p">,</span><span class="w"> </span><span class="n">alt</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">))</span>
<span class="c1"># [1] 80.96026 </span>
<span class="m">-10</span><span class="o">*</span><span class="nf">log10</span><span class="p">(</span><span class="nf">genotype_likelihood</span><span class="p">(</span><span class="n">m</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">g</span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">e</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.01</span><span class="p">,</span><span class="w"> </span><span class="n">ref</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">22</span><span class="p">,</span><span class="w"> </span><span class="n">alt</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">))</span>
<span class="c1"># [1] 78.2678</span>
<span class="m">-10</span><span class="o">*</span><span class="nf">log10</span><span class="p">(</span><span class="nf">genotype_likelihood</span><span class="p">(</span><span class="n">m</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">g</span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">e</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.01</span><span class="p">,</span><span class="w"> </span><span class="n">ref</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">22</span><span class="p">,</span><span class="w"> </span><span class="n">alt</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">))</span>
<span class="c1"># [1] 440.1746</span>
</code></pre></div>
</details>
<p><strong>Exercise:</strong> What is the most likely genotype? What is the genotype quality (GQ)? Do you think we should be confident about this genotype call?</p>
<details class="done">
<summary>Answer</summary>
<p>The most likely genotype has the lowest PL, so where g=1 (heterozygous). GQ is calculated by subtracting the lowest PL from the second lowest PL, so 80.96 - 78.27 = 2.69. </p>
<p>This is a low genotype quality (note that we&rsquo;re in the phred scale), i.e. an error probability of 0.54. This makes sense, if the genotype is heterozygous we would roughly expect to count as many reference as alternative alleles, and our example quite strongly deviates from this expectation. </p>
</details>
<h4 id="calling-variants-with-gatk">Calling variants with GATK</h4>
<p>The command <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037225632-HaplotypeCaller"><code>gatk HaplotypeCaller</code></a> is the core command of <code>gatk</code>. It performs the actual variant calling.</p>
<p><strong>Exercise:</strong> Check out the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037225632-HaplotypeCaller"><code>gatk HaplotypeCaller</code> documentation</a>, and find out which arguments are required.</p>
<details class="done">
<summary>Answer</summary>
<p>Required arguments are:</p>
<ul>
<li><code>--input</code></li>
<li><code>--ouput</code></li>
<li><code>--reference</code></li>
</ul>
</details>
<p><strong>Exercise:</strong> Generate a script called <code>B10_run_haplotype_caller.sh</code> in <code>B-mother_only</code>. Use it to make a directory called <code>~/project/results/variants</code> to write the output vcf. In the same script, run <code>gatk HaplotypeCaller</code> with required options on the recalibrated alignment file of the mother (<code>results/bqsr/mother.recal.bam</code>). We&rsquo;ll focus on a small region, so add <code>--intervals chr20:10018000-10220000</code>.</p>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><span class="filename">B10_run_haplotype_caller.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project
mkdir<span class="w"> </span>-p<span class="w"> </span>results/variants

gatk<span class="w"> </span>HaplotypeCaller<span class="w"> </span><span class="se">\</span>
--reference<span class="w"> </span>data/reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa<span class="w"> </span><span class="se">\</span>
--input<span class="w"> </span>results/bqsr/mother.recal.bam<span class="w"> </span><span class="se">\</span>
--output<span class="w"> </span>results/variants/mother.HC.vcf<span class="w"> </span><span class="se">\</span>
--intervals<span class="w"> </span>chr20:10018000-10220000
</code></pre></div>
</details>
<p><strong>Exercise:</strong> You can get the number of records in a vcf with piping the output of <code>grep -v '^#'</code> to <code>wc -l</code>. Get the number of variants in the vcf.</p>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><pre><span></span><code>grep<span class="w"> </span>-v<span class="w"> </span><span class="s1">&#39;^#&#39;</span><span class="w"> </span>variants/mother.HC.vcf<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l
</code></pre></div>
<p>Shows you that there are 411 variants in there.</p>
</details>
<p>You can get some more statistics with <code>gatk VariantsToTable</code>. The output can be used to easily query things in <code>R</code> or MS Excel.</p>
<p>Here&rsquo;s an example:</p>
<div class="highlight"><pre><span></span><code>gatk<span class="w"> </span>VariantsToTable<span class="w"> </span><span class="se">\</span>
--variant<span class="w"> </span>variants/mother.HC.vcf<span class="w"> </span><span class="se">\</span>
--fields<span class="w"> </span>CHROM<span class="w"> </span>-F<span class="w"> </span>POS<span class="w"> </span>-F<span class="w"> </span>TYPE<span class="w"> </span>-GF<span class="w"> </span>GT<span class="w"> </span><span class="se">\</span>
--output<span class="w"> </span>variants/mother.HC.table
</code></pre></div>
<p><strong>Exercise:</strong> Run the command from within a script called <code>B11_variants_to_table.sh</code>, and have a look at the first few records (use e.g. <code>head</code> or <code>less</code>). After that, report the number of SNPs and INDELs.</p>
<details class="done">
<summary>Answer</summary>
<p>Your script should look like:</p>
<div class="highlight"><span class="filename">B11_variants_to_table.sh</span><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>~/project

gatk<span class="w"> </span>VariantsToTable<span class="w"> </span><span class="se">\</span>
--variant<span class="w"> </span>results/variants/mother.HC.vcf<span class="w"> </span><span class="se">\</span>
--fields<span class="w"> </span>CHROM<span class="w"> </span>-F<span class="w"> </span>POS<span class="w"> </span>-F<span class="w"> </span>TYPE<span class="w"> </span>-GF<span class="w"> </span>GT<span class="w"> </span><span class="se">\</span>
--output<span class="w"> </span>results/variants/mother.HC.table
</code></pre></div>
<p>You can get the number of SNPs with:</p>
<div class="highlight"><pre><span></span><code>grep<span class="w"> </span>-c<span class="w"> </span><span class="s2">&quot;SNP&quot;</span><span class="w"> </span>variants/mother.HC.table
</code></pre></div>
<p>which will give 326</p>
<p>And the number of INDELs with:</p>
<div class="highlight"><pre><span></span><code>grep<span class="w"> </span>-c<span class="w"> </span><span class="s2">&quot;INDEL&quot;</span><span class="w"> </span>variants/mother.HC.table
</code></pre></div>
<p>that outputs 84</p>
<p>A more fancy way to this would be:</p>
<div class="highlight"><pre><span></span><code>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">3</span><span class="w"> </span>variants/mother.HC.table<span class="w"> </span><span class="p">|</span><span class="w"> </span>tail<span class="w"> </span>-n<span class="w"> </span>+2<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span>-c
</code></pre></div>
<p>Giving:</p>
<div class="highlight"><pre><span></span><code>84 INDEL
1 MIXED
326 SNP
</code></pre></div>
</details>
<p>Now, we will perform the variant calling on all three samples. Later we want to combine the variant calls. For efficient merging of vcfs, we will need to output the variants as a GVCF. To do that, we will use the option <code>--emit-ref-confidence GVCF</code>. Also, we&rsquo;ll visualise the haplotype phasing with IGV in the next section. For that we&rsquo;ll need a phased bam. You can get this output with the argument <code>--bam-output</code>.</p>
<p><strong>Exercise:</strong> Create a script in <code>C-all_samples</code> called <code>C06_run_haplotypecaller.sh</code>. Use it to run <code>gatk HaplotypeCaller</code> for mother, father and son in a loop. Use the same arguments as in the previous exercise. On top of that, add the arguments <code>--emit-ref-confidence GVCF</code> and <code>--bamoutput &lt;phased.bam&gt;</code>.</p>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><span class="filename">C06_run_haplotypecaller.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project

<span class="k">for</span><span class="w"> </span>SAMPLE<span class="w"> </span><span class="k">in</span><span class="w"> </span>mother<span class="w"> </span>father<span class="w"> </span>son
<span class="k">do</span>
<span class="w">    </span>gatk<span class="w"> </span>HaplotypeCaller<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--reference<span class="w"> </span>data/reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--input<span class="w"> </span>results/bqsr/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>.recal.bam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--output<span class="w"> </span>results/variants/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>.HC.g.vcf<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--bam-output<span class="w"> </span>results/variants/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>.phased.bam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--intervals<span class="w"> </span>chr20:10018000-10220000<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--emit-ref-confidence<span class="w"> </span>GVCF
<span class="k">done</span>
</code></pre></div>
</details>
<h3 id="2-combining-gvcfs">2. Combining GVCFs</h3>
<p>Now that we have all three GVCFs of the mother, father and son, we can combine them into a database. We do this because it enables us to later add GVCFs (with the option <code>--genomicsdb-update-workspace-path</code>), and to efficiently combine them into a single vcf.</p>
<p>You can generate a GenomicsDB on our three samples like this:</p>
<div class="highlight"><span class="filename">C07_create_genomicsdb.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project

gatk<span class="w"> </span>GenomicsDBImport<span class="w"> </span><span class="se">\</span>
--variant<span class="w"> </span>results/variants/mother.HC.g.vcf<span class="w"> </span><span class="se">\</span>
--variant<span class="w"> </span>results/variants/father.HC.g.vcf<span class="w"> </span><span class="se">\</span>
--variant<span class="w"> </span>results/variants/son.HC.g.vcf<span class="w"> </span><span class="se">\</span>
--intervals<span class="w"> </span>chr20:10018000-10220000<span class="w"> </span><span class="se">\</span>
--genomicsdb-workspace-path<span class="w"> </span>results/genomicsdb
</code></pre></div>
<p><strong>Exercise:</strong> Create a script called <code>C07_create_genomicsdb.sh</code> to run this command to generate the database.</p>
<p>You can retrieve the combined vcf from the database with <code>gatk GenotypeGVCFs</code>.</p>
<div class="highlight"><span class="filename">C08_genotype_gvcfs.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project

gatk<span class="w"> </span>GenotypeGVCFs<span class="w"> </span><span class="se">\</span>
--reference<span class="w"> </span>data/reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa<span class="w"> </span><span class="se">\</span>
--variant<span class="w"> </span>gendb://results/genomicsdb<span class="w"> </span><span class="se">\</span>
--intervals<span class="w"> </span>chr20:10018000-10220000<span class="w"> </span><span class="se">\</span>
--output<span class="w"> </span>results/variants/trio.vcf
</code></pre></div>
<p><strong>Exercise:</strong> Create a script called <code>C08_genotype_gvcfs.sh</code> to run this command to generate the combined vcf.</p>









  




                

  <!-- Giscus -->
  <h2>Questions & Answers</h2>

    <script src="https://giscus.app/client.js"
            data-repo="sib-swiss/NGS-variants-training"
            data-repo-id="MDEwOlJlcG9zaXRvcnkzMTEzMTc0MDI="
            data-category="Q&A"
            data-category-id="DIC_kwDOEo5Tms4CQ1hn"
            data-mapping="pathname"
            data-strict="0"
            data-reactions-enabled="1"
            data-emit-metadata="0"
            data-input-position="top"
            data-theme="preferred_color_scheme"
            data-lang="en"
            crossorigin="anonymous"
            async>
    </script>
  <!-- Synchronize Giscus theme with palette -->
  <script>
    var giscus = document.querySelector("script[src*=giscus]")

    /* Set palette on initial load */
    var palette = __md_get("__palette")
    if (palette && typeof palette.color === "object") {
      var theme = palette.color.scheme === "slate" ? "dark" : "light"
      giscus.setAttribute("data-theme", theme) 


    }

    /* Register event handlers after documented loaded */
    document.addEventListener("DOMContentLoaded", function() {
      var ref = document.querySelector("[data-md-component=palette]")
      ref.addEventListener("change", function() {
        var palette = __md_get("__palette")
        if (palette && typeof palette.color === "object") {
          var theme = palette.color.scheme === "slate" ? "dark" : "light"

          /* Instruct Giscus to change theme */
          var frame = document.querySelector(".giscus-frame")
          frame.contentWindow.postMessage(
            { giscus: { setConfig: { theme } } },
            "https://giscus.app"
          )
        }
      })
    })
  </script>

              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2021-2022 SIB Swiss institute of Bioinformatics –

    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["announce.dismiss"], "search": "../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"default": "latest", "provider": "mike"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.af256bd8.min.js"></script>
      
    
  </body>
</html>