

<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://sib-swiss.github.io/NGS-variants-training/2022.9/day2/variant_calling/">
      
      <link rel="icon" href="../../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.3.0, mkdocs-material-8.2.8">
    
    
      
        <title>Variant calling - NGS - variant analysis</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.644de097.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.e6a45f82.min.css">
        
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-SK9D4D2P4M"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-SK9D4D2P4M');
</script>

    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#learning-outcomes" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-component="outdated" hidden>
        <aside class="md-banner md-banner--warning">
          
        </aside>
      </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="NGS - variant analysis" class="md-header__button md-logo" aria-label="NGS - variant analysis" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            NGS - variant analysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Variant calling
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/sib-swiss/NGS-variants-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/NGS-variants-training
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="NGS - variant analysis" class="md-nav__button md-logo" aria-label="NGS - variant analysis" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    NGS - variant analysis
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/sib-swiss/NGS-variants-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/NGS-variants-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../precourse/" class="md-nav__link">
        Precourse preparations
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../course_schedule/" class="md-nav__link">
        Course schedule
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4">
          Day 1
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Day 1" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Day 1
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/introduction/" class="md-nav__link">
        Introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/server_login/" class="md-nav__link">
        Setup
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/reproducibility/" class="md-nav__link">
        Reproducibility
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/alignment/" class="md-nav__link">
        Read alignment - basics
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/alignment_advanced/" class="md-nav__link">
        Read alignment - advanced
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          Day 2
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Day 2" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          Day 2
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../variant_calling_preparation/" class="md-nav__link">
        Variant calling - preparation
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Variant calling
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Variant calling
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    Learning outcomes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    Material
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    Exercises
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-variant-calling" class="md-nav__link">
    1. Variant calling
  </a>
  
    <nav class="md-nav" aria-label="1. Variant calling">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#calculating-pl-and-gq-by-hand" class="md-nav__link">
    Calculating PL and GQ by hand
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#calling-variants-with-gatk" class="md-nav__link">
    Calling variants with GATK
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-combining-gvcfs" class="md-nav__link">
    2. Combining GVCFs
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../visualisation/" class="md-nav__link">
        Visualisation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../filtering_evaluation/" class="md-nav__link">
        Filtering & evaluation
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../annotation/" class="md-nav__link">
        Annotation
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    Learning outcomes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    Material
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    Exercises
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-variant-calling" class="md-nav__link">
    1. Variant calling
  </a>
  
    <nav class="md-nav" aria-label="1. Variant calling">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#calculating-pl-and-gq-by-hand" class="md-nav__link">
    Calculating PL and GQ by hand
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#calling-variants-with-gatk" class="md-nav__link">
    Calling variants with GATK
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-combining-gvcfs" class="md-nav__link">
    2. Combining GVCFs
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
<a href="https://github.com/sib-swiss/NGS-variants-training/edit/main/docs/day2/variant_calling.md" title="Edit this page" class="md-content__button md-icon">
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
</a>



  <h1>Variant calling</h1>

<h2 id="learning-outcomes">Learning outcomes</h2>
<p><strong>After having completed this chapter you will be able to:</strong></p>
<ul>
<li>Perform basic calculations regarding the genotype likelihood of individual variants</li>
<li>Follow <code>gatk</code> best practices workflow to perform a variant analysis by:<ul>
<li>Calling variants with <code>gatk HaplotypeCaller</code></li>
<li>Combining multiple <code>vcf</code> files into a single <code>vcf</code> file</li>
</ul>
</li>
<li>Perform basic operations to get statistics of a <code>vcf</code> file</li>
</ul>
<h2 id="material">Material</h2>
<p>The <a href="https://www.biorxiv.org/content/10.1101/201178v3">paper</a> on genomic variant call format (gVCF)</p>
<p><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035535932-Germline-short-variant-discovery-SNPs-Indels-">GATK best practices germline short variant workflow</a>:</p>
<figure>
  <img src="../../assets/images/gatk_germline.png" width="800"/>
</figure>

<h2 id="exercises">Exercises</h2>
<h3 id="1-variant-calling">1. Variant calling</h3>
<h4 id="calculating-pl-and-gq-by-hand">Calculating PL and GQ by hand</h4>
<p>Here&rsquo;s a function in R to calculate genotype likelihoods as described in Li H.  Bioinformatics. 2011;27:2987â€“93 (assuming equal base error probabilities for all reads):</p>
<div class="highlight"><pre><span></span><code><span class="n">genotype_likelihood</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">m</span><span class="p">,</span><span class="n">g</span><span class="p">,</span><span class="n">e</span><span class="p">,</span><span class="n">ref</span><span class="p">,</span><span class="n">alt</span><span class="p">){</span>
  <span class="p">(((</span><span class="n">m</span><span class="o">-</span><span class="n">g</span><span class="p">)</span><span class="o">*</span><span class="n">e</span><span class="o">+</span><span class="n">g</span><span class="o">*</span><span class="p">(</span><span class="m">1</span><span class="o">-</span><span class="n">e</span><span class="p">))</span><span class="o">^</span><span class="n">alt</span> <span class="o">*</span> <span class="p">((</span><span class="n">m</span><span class="o">-</span><span class="n">g</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="m">1</span><span class="o">-</span><span class="n">e</span><span class="p">)</span><span class="o">+</span><span class="n">g</span><span class="o">*</span><span class="n">e</span><span class="p">)</span><span class="o">^</span><span class="n">ref</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">m</span><span class="o">^</span><span class="p">(</span><span class="n">ref</span><span class="o">+</span><span class="n">alt</span><span class="p">))</span>
<span class="p">}</span>
</code></pre></div>
<p>Where:</p>
<ul>
<li><code>m</code> : ploidy</li>
<li><code>g</code> : number of alternative alleles</li>
<li><code>e</code> : base error probability</li>
<li><code>ref</code> : number of reference alleles counted</li>
<li><code>alt</code> : number of alternative alleles counted</li>
</ul>
<p><strong>Exercise:</strong> In a local R session, calculate the three genotype likelihoods (for g = 0, g = 1 and g = 2) for a case where we count 22 reference alleles and 4 alternative alleles (so a coverage of 26), and base error probability of 0.01. Calculate the PL values (<code>-10*log10(likelihood)</code>) for each genotype. </p>
<div class="admonition note">
<p class="admonition-title">No local R installation?</p>
<p>If you don&rsquo;t have access to an R installation, you can also install it in the jupyter environment by running inside the terminal:</p>
<div class="highlight"><pre><span></span><code>conda install r-base 
</code></pre></div>
<p>After installation, start an interactive R session by typing <code>R</code>. </p>
</div>
<details class="done"><summary>Answer</summary><div class="highlight"><pre><span></span><code><span class="c1"># For g = 0 (i.e. 0 reference alleles)</span>
<span class="m">-10</span><span class="o">*</span><span class="nf">log10</span><span class="p">(</span><span class="nf">genotype_likelihood</span><span class="p">(</span><span class="n">m</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span> <span class="n">g</span><span class="o">=</span> <span class="m">0</span><span class="p">,</span> <span class="n">e</span> <span class="o">=</span> <span class="m">0.01</span><span class="p">,</span> <span class="n">ref</span> <span class="o">=</span> <span class="m">22</span><span class="p">,</span> <span class="n">alt</span> <span class="o">=</span> <span class="m">4</span><span class="p">))</span>
<span class="c1"># [1] 80.96026 </span>
<span class="m">-10</span><span class="o">*</span><span class="nf">log10</span><span class="p">(</span><span class="nf">genotype_likelihood</span><span class="p">(</span><span class="n">m</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span> <span class="n">g</span><span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">e</span> <span class="o">=</span> <span class="m">0.01</span><span class="p">,</span> <span class="n">ref</span> <span class="o">=</span> <span class="m">22</span><span class="p">,</span> <span class="n">alt</span> <span class="o">=</span> <span class="m">4</span><span class="p">))</span>
<span class="c1"># [1] 78.2678</span>
<span class="m">-10</span><span class="o">*</span><span class="nf">log10</span><span class="p">(</span><span class="nf">genotype_likelihood</span><span class="p">(</span><span class="n">m</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span> <span class="n">g</span><span class="o">=</span> <span class="m">2</span><span class="p">,</span> <span class="n">e</span> <span class="o">=</span> <span class="m">0.01</span><span class="p">,</span> <span class="n">ref</span> <span class="o">=</span> <span class="m">22</span><span class="p">,</span> <span class="n">alt</span> <span class="o">=</span> <span class="m">4</span><span class="p">))</span>
<span class="c1"># [1] 440.1746</span>
</code></pre></div>
</details>
<p><strong>Exercise:</strong> What is the most likely genotype? What is the genotype quality (GQ)? Do you think we should be confident about this genotype call?</p>
<details class="done"><summary>Answer</summary><p>The most likely genotype has the lowest PL, so where g=1 (heterozygous). GL is calculated by subtracting the lowest PL from the second lowest PL, so 80.96 - 78.27 = 2.69. </p>
<p>This is a low genotype quality (note that we&rsquo;re in the phred scale), i.e. an error probability of 0.54. This makes sense, if the genotype is heterozygous we would roughly expect to count as many reference as alternative alleles, and our example quite strongly deviates from this expectation. </p>
</details>
<h4 id="calling-variants-with-gatk">Calling variants with GATK</h4>
<p>The command <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037225632-HaplotypeCaller"><code>gatk HaplotypeCaller</code></a> is the core command of <code>gatk</code>. It performs the actual variant calling.</p>
<p><strong>Exercise:</strong> Check out the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037225632-HaplotypeCaller"><code>gatk HaplotypeCaller</code> documentation</a>, and find out which arguments are required.</p>
<details class="done"><summary>Answer</summary><p>Required arguments are:</p>
<ul>
<li><code>--input</code></li>
<li><code>--ouput</code></li>
<li><code>--reference</code></li>
</ul>
</details>
<p><strong>Exercise:</strong> Generate a script called <code>B10_run_haplotype_caller.sh</code> in <code>B-mother_only</code>. Use it to make a directory called <code>~/workdir/results/variants</code> to write the output vcf. In the same script, run <code>gatk HaplotypeCaller</code> with required options on the recalibrated alignment file of the mother (<code>results/bqsr/mother.recal.bam</code>). We&rsquo;ll focus on a small region, so add <code>--intervals chr20:10018000-10220000</code>.</p>
<details class="done"><summary>Answer</summary><div class="highlight"><span class="filename">B10_run_haplotype_caller.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span> ~/workdir
mkdir -p results/variants

gatk HaplotypeCaller <span class="se">\</span>
--reference data/reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa <span class="se">\</span>
--input results/bqsr/mother.recal.bam <span class="se">\</span>
--output results/variants/mother.HC.vcf <span class="se">\</span>
--intervals chr20:10018000-10220000
</code></pre></div>
</details>
<p><strong>Exercise:</strong> You can get the number of records in a vcf with piping the output of <code>grep -v '^#'</code> to <code>wc -l</code>. Get the number of variants in the vcf.</p>
<details class="done"><summary>Answer</summary><div class="highlight"><pre><span></span><code>grep -v <span class="s1">&#39;^#&#39;</span> variants/mother.HC.vcf <span class="p">|</span> wc -l
</code></pre></div>
<p>Shows you that there are 411 variants in there.</p>
</details>
<p>You can get some more statistics with <code>gatk VariantsToTable</code>. The output can be used to easily query things in <code>R</code> or MS Excel.</p>
<p>Here&rsquo;s an example:</p>
<div class="highlight"><pre><span></span><code>gatk VariantsToTable <span class="se">\</span>
--variant variants/mother.HC.vcf <span class="se">\</span>
--fields CHROM -F POS -F TYPE -GF GT <span class="se">\</span>
--output variants/mother.HC.table
</code></pre></div>
<p><strong>Exercise:</strong> Run the command from within a script called <code>B11_variants_to_table.sh</code>, and have a look at the first few records (use e.g. <code>head</code> or <code>less</code>). After that, report the number of SNPs and INDELs.</p>
<details class="done"><summary>Answer</summary><p>Your script should look like:</p>
<div class="highlight"><span class="filename">B11_variants_to_table.sh</span><pre><span></span><code><span class="nb">cd</span> ~/workdir

gatk VariantsToTable <span class="se">\</span>
--variant results/variants/mother.HC.vcf <span class="se">\</span>
--fields CHROM -F POS -F TYPE -GF GT <span class="se">\</span>
--output results/variants/mother.HC.table
</code></pre></div>
<p>You can get the number of SNPs with:</p>
<div class="highlight"><pre><span></span><code>grep -c <span class="s2">&quot;SNP&quot;</span> variants/mother.HC.table
</code></pre></div>
<p>which will give 326</p>
<p>And the number of INDELs with:</p>
<div class="highlight"><pre><span></span><code>grep -c <span class="s2">&quot;INDEL&quot;</span> variants/mother.HC.table
</code></pre></div>
<p>that outputs 84</p>
<p>A more fancy way to this would be:</p>
<div class="highlight"><pre><span></span><code>cut -f <span class="m">3</span> variants/mother.HC.table <span class="p">|</span> tail -n +2 <span class="p">|</span> sort <span class="p">|</span> uniq -c
</code></pre></div>
<p>Giving:</p>
<div class="highlight"><pre><span></span><code>84 INDEL
1 MIXED
326 SNP
</code></pre></div>
</details>
<p>Now, we will perform the variant calling on all three samples. Later we want to combine the variant calls. For efficient merging of vcfs, we will need to output the variants as a GVCF. To do that, we will use the option <code>--emit-ref-confidence GVCF</code>. Also, we&rsquo;ll visualise the haplotype phasing with IGV in the next section. For that we&rsquo;ll need a phased bam. You can get this output with the argument <code>--bam-output</code>.</p>
<p><strong>Exercise:</strong> Create a script in <code>C-all_samples</code> called <code>C06_run_haplotypecaller.sh</code>. Use it to run <code>gatk HaplotypeCaller</code> for mother, father and son in a loop. Use the same arguments as in the previous exercise. On top of that, add the arguments <code>--emit-ref-confidence GVCF</code> and <code>--bamoutput &lt;phased.bam&gt;</code>.</p>
<details class="done"><summary>Answer</summary><div class="highlight"><span class="filename">C06_run_haplotypecaller.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span> ~/workdir

<span class="k">for</span> SAMPLE <span class="k">in</span> mother father son
<span class="k">do</span>
    gatk HaplotypeCaller <span class="se">\</span>
    --reference data/reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa <span class="se">\</span>
    --input results/bqsr/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>.recal.bam <span class="se">\</span>
    --output results/variants/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>.HC.g.vcf <span class="se">\</span>
    --bam-output results/variants/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>.phased.bam <span class="se">\</span>
    --intervals chr20:10018000-10220000 <span class="se">\</span>
    --emit-ref-confidence GVCF
<span class="k">done</span>
</code></pre></div>
</details>
<h3 id="2-combining-gvcfs">2. Combining GVCFs</h3>
<p>Now that we have all three GVCFs of the mother, father and son, we can combine them into a database. We do this because it enables us to later add GVCFs (with the option <code>--genomicsdb-update-workspace-path</code>), and to efficiently combine them into a single vcf.</p>
<p>You can generate a GenomicsDB on our three samples like this:</p>
<div class="highlight"><span class="filename">C07_create_genomicsdb.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span> ~/workdir

gatk GenomicsDBImport <span class="se">\</span>
--variant results/variants/mother.HC.g.vcf <span class="se">\</span>
--variant results/variants/father.HC.g.vcf <span class="se">\</span>
--variant results/variants/son.HC.g.vcf <span class="se">\</span>
--intervals chr20:10018000-10220000 <span class="se">\</span>
--genomicsdb-workspace-path results/genomicsdb
</code></pre></div>
<p><strong>Exercise:</strong> Create a script called <code>C07_create_genomicsdb.sh</code> to run this command to generate the database.</p>
<p>You can retrieve the combined vcf from the database with <code>gatk GenotypeGVCFs</code>.</p>
<div class="highlight"><span class="filename">C08_genotype_gvcfs.sh</span><pre><span></span><code>gatk GenotypeGVCFs <span class="se">\</span>
--reference data/reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa <span class="se">\</span>
--variant gendb://genomicsdb <span class="se">\</span>
--intervals chr20:10018000-10220000 <span class="se">\</span>
--output variants/trio.vcf
</code></pre></div>
<p><strong>Exercise:</strong> Create a script called <code>C08_genotype_gvcfs.sh</code> to run this command to generate the combined vcf.</p>

              
            </article>
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    <nav class="md-footer__inner md-grid" aria-label="Footer">
      
        
        <a href="../variant_calling_preparation/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Variant calling - preparation" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Variant calling - preparation
            </div>
          </div>
        </a>
      
      
        
        <a href="../visualisation/" class="md-footer__link md-footer__link--next" aria-label="Next: Visualisation" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Visualisation
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2021 SIB Swiss institute of Bioinformatics
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../..", "features": [], "search": "../../assets/javascripts/workers/search.5e67fbfe.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.c44cc438.min.js"></script>
      
    
  </body>
</html>