
<!doctype html>
<html lang="en" class="no-js">
  <head>
    

      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://sib-swiss.github.io/NGS-variants-training/2024.9/day2/filtering_evaluation/">
      
      
        <link rel="prev" href="../visualisation/">
      
      
        <link rel="next" href="../annotation/">
      
      
      <link rel="icon" href="../../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.33">
    
<script type="application/ld+json">
  {
    "@context": "https://schema.org/",
    "@type": "LearningResource",
    "@id": "https://sib-swiss.github.io/NGS-variants-training",
    "http://purl.org/dc/terms/conformsTo": {
        "@type": "CreativeWork",
        "@id": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE"
    },
    "description": "Course on detecting variants from next generation sequencing data",
    "keywords": "NGS, variant, SNP, INDEL, sequencing, SNV, gatk",
    "name": "NGS - variant analysis",
    "about": [
        {
            "@type": "DefinedTerm",
            "@id": "http://edamontology.org/topic_3168",
            "inDefinedTermSet": "http://edamontology.org",
            "termCode": "topic_3168",
            "name": "Sequencing"
        },
        {
            "@type": "DefinedTerm",
            "@id": "http://edamontology.org/topic_3227",
            "inDefinedTermSet": "http://edamontology.org",
            "termCode": "topic_3227",
            "name": "Variant calling"
        }
    ],
    "abstract": "The detection of genetic variation is of major interest in various disciplines spanning from ecology and evolution research to inherited disease discovery and precision oncology. Next generation sequencing (NGS) methods are very powerful for the detection of genomic variants. Thanks to its throughput and cost-efficiency it enables the detection of a large number of variants in a large number of samples. In this two-day course we will cover the steps from read alignment to variant calling and annotation. We will mainly focus on the detection of germline mutations by following the GATK best practices. ",
    "audience": "Biologists working with NGS data",
    "author": [
        {
            "@type": "Person",
            "name": "Geert van Geest",
            "email": "geert.vangeest@sib.swiss"
        },
        {
            "@type": "Person",
            "name": "Patricia Palagi"
        },
        {
            "@type": "Organization",
            "name": "SIB Swiss Institute of Bioinformatics"
        }
    ],
    "identifier": "https://doi.org/10.5281/zenodo.6457818",
    "license": "https://creativecommons.org/licenses/by/4.0/",
    "teaches": [
        "Understand important aspects of NGS and read alignment for variant analysis",
        "Perform a read alignment ready for variant analysis",
        "Perform variant calling according to GATK best practices",
        "Perform a variant annotation"
    ]
}
</script>

    
      
        <title>Filtering & evaluation - NGS - variant analysis</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.3cba04c6.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
<!-- Matomo -->
<script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
      var u="https://matomo.sib.swiss/";
      _paq.push(['setTrackerUrl', u+'matomo.php']);
      _paq.push(['setSiteId', '220']);
      var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
      g.async=true; g.src=u+'matomo.js'; s.parentNode.insertBefore(g,s);
    })();
  </script>
  <!-- End Matomo Code -->
  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#learning-outcomes" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
        <aside class="md-banner">
          <div class="md-banner__inner md-grid md-typeset">
            
              <button class="md-banner__button md-icon" aria-label="Don't show this again">
                
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
              </button>
            
            
  <b>Issues, questions or feedback? &#128591; &#128640; </b> Use the "Questions and Answers" section at the bottom &#128071; of every page

          </div>
          
            <script>var content,el=document.querySelector("[data-md-component=announce]");el&&(content=el.querySelector(".md-typeset"),__md_hash(content.innerHTML)===__md_get("__announce")&&(el.hidden=!0))</script>
          
        </aside>
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
          <aside class="md-banner md-banner--warning">
            <div class="md-banner__inner md-grid md-typeset">
              
  You're not viewing the latest version.
  <a href="../../.."> 
    <strong>Click here to go to latest.</strong>
  </a>

            </div>
            <script>var el=document.querySelector("[data-md-component=outdated]"),outdated=__md_get("__outdated",sessionStorage);!0===outdated&&el&&(el.hidden=!1)</script>
          </aside>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="NGS - variant analysis" class="md-header__button md-logo" aria-label="NGS - variant analysis" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            NGS - variant analysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Filtering & evaluation
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/sib-swiss/NGS-variants-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/NGS-variants-training
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="NGS - variant analysis" class="md-nav__button md-logo" aria-label="NGS - variant analysis" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    NGS - variant analysis
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/sib-swiss/NGS-variants-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/NGS-variants-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../precourse/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Precourse preparations
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../course_schedule/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Course schedule
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Day 1
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Day 1
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/server_login/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Setup
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/reproducibility/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Reproducibility
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/alignment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Read alignment - basics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/alignment_advanced/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Read alignment - advanced
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Day 2
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Day 2
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variant_calling_preparation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant calling - preparation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variant_calling/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant calling
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../visualisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Visualisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Filtering & evaluation
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Filtering & evaluation
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    <span class="md-ellipsis">
      Learning outcomes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    <span class="md-ellipsis">
      Material
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    <span class="md-ellipsis">
      Exercises
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-hard-filtering" class="md-nav__link">
    <span class="md-ellipsis">
      1. Hard filtering
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Hard filtering">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#splitting-snps-and-indels" class="md-nav__link">
    <span class="md-ellipsis">
      Splitting SNPs and INDELs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering-snps" class="md-nav__link">
    <span class="md-ellipsis">
      Filtering SNPs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering-indels" class="md-nav__link">
    <span class="md-ellipsis">
      Filtering INDELs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#merging-filtered-snps-and-indels" class="md-nav__link">
    <span class="md-ellipsis">
      Merging filtered SNPs and INDELs
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-evaluation-by-concordance" class="md-nav__link">
    <span class="md-ellipsis">
      2. Evaluation by concordance
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Annotation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    <span class="md-ellipsis">
      Learning outcomes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    <span class="md-ellipsis">
      Material
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    <span class="md-ellipsis">
      Exercises
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-hard-filtering" class="md-nav__link">
    <span class="md-ellipsis">
      1. Hard filtering
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Hard filtering">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#splitting-snps-and-indels" class="md-nav__link">
    <span class="md-ellipsis">
      Splitting SNPs and INDELs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering-snps" class="md-nav__link">
    <span class="md-ellipsis">
      Filtering SNPs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering-indels" class="md-nav__link">
    <span class="md-ellipsis">
      Filtering INDELs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#merging-filtered-snps-and-indels" class="md-nav__link">
    <span class="md-ellipsis">
      Merging filtered SNPs and INDELs
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-evaluation-by-concordance" class="md-nav__link">
    <span class="md-ellipsis">
      2. Evaluation by concordance
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
  
                  

  
  


  <h1>Filtering & evaluation</h1>

<h2 id="learning-outcomes">Learning outcomes</h2>
<p><strong>After having completed this chapter you will be able to:</strong></p>
<ul>
<li>Explain why using Variant Quality Score Recalibration (VQSR) for filtering variants can outperform hard filtering </li>
<li>Perform hard filtering on both SNPs and INDELs separately by using <code>gatk SelectVariants</code> in combination with <code>gatk VariantFiltration</code></li>
<li>Perform concordance between called variants and a truth set and evaluate performance of a variant calling workflow</li>
</ul>
<h2 id="material">Material</h2>
<iframe width="560" height="315" src="https://www.youtube.com/embed/8oiQOg40_co?si=bCnSYkIrWB_yCjjN" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

<p><a class="md-button" href="../../assets/pdf/06_filtering_evaluation.pdf"><span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M0 64C0 28.7 28.7 0 64 0h160v128c0 17.7 14.3 32 32 32h128v144H176c-35.3 0-64 28.7-64 64v144H64c-35.3 0-64-28.7-64-64V64zm384 64H256V0l128 128zM176 352h32c30.9 0 56 25.1 56 56s-25.1 56-56 56h-16v32c0 8.8-7.2 16-16 16s-16-7.2-16-16V368c0-8.8 7.2-16 16-16zm32 80c13.3 0 24-10.7 24-24s-10.7-24-24-24h-16v48h16zm96-80h32c26.5 0 48 21.5 48 48v64c0 26.5-21.5 48-48 48h-32c-8.8 0-16-7.2-16-16V368c0-8.8 7.2-16 16-16zm32 128c8.8 0 16-7.2 16-16v-64c0-8.8-7.2-16-16-16h-16v96h16zm80-112c0-8.8 7.2-16 16-16h48c8.8 0 16 7.2 16 16s-7.2 16-16 16h-32v32h32c8.8 0 16 7.2 16 16s-7.2 16-16 16h-32v48c0 8.8-7.2 16-16 16s-16-7.2-16-16V368z"/></svg></span> Download the presentation</a></p>
<h2 id="exercises">Exercises</h2>
<h3 id="1-hard-filtering">1. Hard filtering</h3>
<p>The developers of <code>gatk</code> strongly advise to do <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531612-Variant-Quality-Score-Recalibration-VQSR-">Variant Quality Score Recalibration (VQSR)</a> for filtering SNPs and INDELs. However, this is not always possible. For example, in the case of limited data availability and/or in the case you are working with non-model organisms and/or in the case you are a bit lazy and okay with a number of false positives.</p>
<p>Our dataset is too small to apply VQSR. We will therefore do hard filtering instead.</p>
<h4 id="splitting-snps-and-indels">Splitting SNPs and INDELs</h4>
<p>First, filtering thresholds are usually different for SNPs and INDELs. Therefore, we will split <code>trio.vcf</code> into two vcfs, one containg only SNPs, and one containing only INDELs. You can extract all the SNP records in our trio vcf like this:</p>
<div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>~/project/results

gatk<span class="w"> </span>SelectVariants<span class="w"> </span><span class="se">\</span>
--variant<span class="w"> </span>variants/trio.vcf<span class="w"> </span><span class="se">\</span>
--select-type-to-include<span class="w"> </span>SNP<span class="w"> </span><span class="se">\</span>
--output<span class="w"> </span>variants/trio.SNP.vcf
</code></pre></div>
<p><strong>Exercise:</strong> Check out the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037055952-SelectVariants">documentation of <code>gatk SelectVariants</code></a>, and:</p>
<ul>
<li>Figure out what you&rsquo;ll need to write at <code>--select-type-to-include</code> if you want to select only INDELS.</li>
<li>Make a script (named <code>C09_select_SNPs.sh</code>) to generate a vcf with only the SNPs</li>
<li>Make a script (named <code>C10_select_INDELs.sh</code>) to generate a second vcf with only the INDELs from <code>trio.vcf</code>.</li>
</ul>
<details class="done">
<summary>Answer</summary>
<p>You will need to fill in <code>INDEL</code> at <code>--select-type-to-include</code> to filter for INDELs.</p>
<p>To get the SNPs you can run the command above:</p>
<div class="highlight"><span class="filename">C09_select_SNPs.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project

gatk<span class="w"> </span>SelectVariants<span class="w"> </span><span class="se">\</span>
--variant<span class="w"> </span>results/variants/trio.vcf<span class="w"> </span><span class="se">\</span>
--select-type-to-include<span class="w"> </span>SNP<span class="w"> </span><span class="se">\</span>
--output<span class="w"> </span>results/variants/trio.SNP.vcf
</code></pre></div>
<p>To get the INDELs you&rsquo;ll need to change <code>--select-type-to-include</code> to <code>INDEL</code>:</p>
<div class="highlight"><span class="filename">C10_select_INDELs.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project

gatk<span class="w"> </span>SelectVariants<span class="w"> </span><span class="se">\</span>
--variant<span class="w"> </span>results/variants/trio.vcf<span class="w"> </span><span class="se">\</span>
--select-type-to-include<span class="w"> </span>INDEL<span class="w"> </span><span class="se">\</span>
--output<span class="w"> </span>results/variants/trio.INDEL.vcf
</code></pre></div>
</details>
<h4 id="filtering-snps">Filtering SNPs</h4>
<p>The command <code>gatk VariantFiltration</code> enables you to filter for both the INFO field (per variant) and FORMAT field (per genotype). For now we&rsquo;re only interested in filtering variants. Below you can find the command to hard-filter the SNP variants on some sensible thresholds (that are explained <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890471-Hard-filtering-germline-short-variants">here</a>).</p>
<div class="highlight"><pre><span></span><code>gatk<span class="w"> </span>VariantFiltration<span class="w"> </span><span class="se">\</span>
--variant<span class="w"> </span>variants/trio.SNP.vcf<span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;QD &lt; 2.0&quot;</span><span class="w">              </span>--filter-name<span class="w"> </span><span class="s2">&quot;QD2&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;QUAL &lt; 30.0&quot;</span><span class="w">           </span>--filter-name<span class="w"> </span><span class="s2">&quot;QUAL30&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;SOR &gt; 3.0&quot;</span><span class="w">             </span>--filter-name<span class="w"> </span><span class="s2">&quot;SOR3&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;FS &gt; 60.0&quot;</span><span class="w">             </span>--filter-name<span class="w"> </span><span class="s2">&quot;FS60&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;MQ &lt; 40.0&quot;</span><span class="w">             </span>--filter-name<span class="w"> </span><span class="s2">&quot;MQ40&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;MQRankSum &lt; -12.5&quot;</span><span class="w">     </span>--filter-name<span class="w"> </span><span class="s2">&quot;MQRankSum-12.5&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;ReadPosRankSum &lt; -8.0&quot;</span><span class="w"> </span>--filter-name<span class="w"> </span><span class="s2">&quot;ReadPosRankSum-8&quot;</span><span class="w"> </span><span class="se">\</span>
--output<span class="w"> </span>variants/trio.SNP.filtered.vcf
</code></pre></div>
<p><strong>Exercise:</strong> Run the filtering command above in a script called <code>C11_filter_SNPs.sh</code>. Did it affect the number of records in the vcf?</p>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>You can check out the number of records in a vcf with:</p>
<div class="highlight"><pre><span></span><code>grep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;^#&quot;</span><span class="w"> </span>&lt;variants.vcf&gt;<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l
</code></pre></div>
</div>
<details class="done">
<summary>Answer</summary>
<p>Your script:</p>
<div class="highlight"><span class="filename">C11_filter_SNPs.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project/results/variants

gatk<span class="w"> </span>VariantFiltration<span class="w"> </span><span class="se">\</span>
--variant<span class="w"> </span>trio.SNP.vcf<span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;QD &lt; 2.0&quot;</span><span class="w">              </span>--filter-name<span class="w"> </span><span class="s2">&quot;QD2&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;QUAL &lt; 30.0&quot;</span><span class="w">           </span>--filter-name<span class="w"> </span><span class="s2">&quot;QUAL30&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;SOR &gt; 3.0&quot;</span><span class="w">             </span>--filter-name<span class="w"> </span><span class="s2">&quot;SOR3&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;FS &gt; 60.0&quot;</span><span class="w">             </span>--filter-name<span class="w"> </span><span class="s2">&quot;FS60&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;MQ &lt; 40.0&quot;</span><span class="w">             </span>--filter-name<span class="w"> </span><span class="s2">&quot;MQ40&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;MQRankSum &lt; -12.5&quot;</span><span class="w">     </span>--filter-name<span class="w"> </span><span class="s2">&quot;MQRankSum-12.5&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;ReadPosRankSum &lt; -8.0&quot;</span><span class="w"> </span>--filter-name<span class="w"> </span><span class="s2">&quot;ReadPosRankSum-8&quot;</span><span class="w"> </span><span class="se">\</span>
--output<span class="w"> </span>trio.SNP.filtered.vcf
</code></pre></div>
<p>There are no differences in the number of records:</p>
<div class="highlight"><pre><span></span><code>grep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;^#&quot;</span><span class="w"> </span>variants/trio.SNP.vcf<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l
</code></pre></div>
<p>and</p>
<div class="highlight"><pre><span></span><code>grep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;^#&quot;</span><span class="w"> </span>variants/trio.SNP.filtered.vcf<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l
</code></pre></div>
<p>both give 446.</p>
<p>However, there are SNPs filtered out, by changing the <code>FILTER</code> column. You can check the number of records with PASS by:</p>
<div class="highlight"><pre><span></span><code>grep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;^#&quot;</span><span class="w"> </span>variants/trio.SNP.filtered.vcf<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">7</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span>-c
</code></pre></div>
<p>Giving:</p>
<div class="highlight"><pre><span></span><code>441 PASS
2 QD2;SOR3
3 SOR3
</code></pre></div>
</details>
<h4 id="filtering-indels">Filtering INDELs</h4>
<p>A command with <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531112--How-to-Filter-variants-either-with-VQSR-or-by-hard-filtering">sensible parameters</a> to do a first iteration of hard filtering the INDELs would be:</p>
<div class="highlight"><pre><span></span><code>gatk<span class="w"> </span>VariantFiltration<span class="w"> </span><span class="se">\</span>
--variant<span class="w"> </span>variants/trio.INDEL.vcf<span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;QD &lt; 2.0&quot;</span><span class="w">                  </span>--filter-name<span class="w"> </span><span class="s2">&quot;QD2&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;QUAL &lt; 30.0&quot;</span><span class="w">               </span>--filter-name<span class="w"> </span><span class="s2">&quot;QUAL30&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;FS &gt; 200.0&quot;</span><span class="w">                </span>--filter-name<span class="w"> </span><span class="s2">&quot;FS200&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;ReadPosRankSum &lt; -20.0&quot;</span><span class="w">    </span>--filter-name<span class="w"> </span><span class="s2">&quot;ReadPosRankSum-20&quot;</span><span class="w"> </span><span class="se">\</span>
--output<span class="w"> </span>variants/trio.INDEL.filtered.vcf
</code></pre></div>
<p><strong>Exercise:</strong> Run the command from a script called <code>C12_filter_INDELs.sh</code> and figure out how many variants are filtered out.</p>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>You can use this command from the answer to the previous exercise:</p>
<div class="highlight"><pre><span></span><code>grep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;^#&quot;</span><span class="w"> </span>&lt;variants.vcf&gt;<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">7</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span>-c
</code></pre></div>
<p>to see how many INDELs were filtered out.</p>
</div>
<details class="done">
<summary>Answer</summary>
<p>Your script:</p>
<div class="highlight"><span class="filename">C12_filter_INDELs.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project/results/variants

gatk<span class="w"> </span>VariantFiltration<span class="w"> </span><span class="se">\</span>
--variant<span class="w"> </span>trio.INDEL.vcf<span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;QD &lt; 2.0&quot;</span><span class="w">                  </span>--filter-name<span class="w"> </span><span class="s2">&quot;QD2&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;QUAL &lt; 30.0&quot;</span><span class="w">               </span>--filter-name<span class="w"> </span><span class="s2">&quot;QUAL30&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;FS &gt; 200.0&quot;</span><span class="w">                </span>--filter-name<span class="w"> </span><span class="s2">&quot;FS200&quot;</span><span class="w"> </span><span class="se">\</span>
--filter-expression<span class="w"> </span><span class="s2">&quot;ReadPosRankSum &lt; -20.0&quot;</span><span class="w">    </span>--filter-name<span class="w"> </span><span class="s2">&quot;ReadPosRankSum-20&quot;</span><span class="w"> </span><span class="se">\</span>
--output<span class="w"> </span>trio.INDEL.filtered.vcf
</code></pre></div>
<p>And check out the contents of the <code>FILTER</code> column: </p>
<div class="highlight"><pre><span></span><code>grep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;^#&quot;</span><span class="w"> </span>variants/trio.INDEL.filtered.vcf<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">7</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span>-c
</code></pre></div>
<p>gives:</p>
<div class="highlight"><pre><span></span><code>110 PASS
</code></pre></div>
<p>So no variants are filtered out.</p>
</details>
<h4 id="merging-filtered-snps-and-indels">Merging filtered SNPs and INDELs</h4>
<p>Now that we have filtered the INDELs and SNPs separately, we can merge them again with this command:</p>
<div class="highlight"><pre><span></span><code>gatk<span class="w"> </span>MergeVcfs<span class="w"> </span><span class="se">\</span>
--INPUT<span class="w"> </span>&lt;input1.vcf&gt;<span class="w"> </span><span class="se">\</span>
--INPUT<span class="w"> </span>&lt;input2.vcf&gt;<span class="w"> </span><span class="se">\</span>
--OUTPUT<span class="w"> </span>&lt;merged.vcf&gt;
</code></pre></div>
<p><strong>Exercise:</strong> Run this command from a script called <code>C13_merge_filtered.sh</code> to merge the vcfs (<code>trio.SNP.filtered.vcf</code> and <code>trio.INDEL.filtered.vcf</code>).</p>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><span class="filename">C13_merged_filtered.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project/results/variants

gatk<span class="w"> </span>MergeVcfs<span class="w"> </span><span class="se">\</span>
--INPUT<span class="w"> </span>trio.SNP.filtered.vcf<span class="w"> </span><span class="se">\</span>
--INPUT<span class="w"> </span>trio.INDEL.filtered.vcf<span class="w"> </span><span class="se">\</span>
--OUTPUT<span class="w"> </span>trio.filtered.vcf
</code></pre></div>
</details>
<h3 id="2-evaluation-by-concordance">2. Evaluation by concordance</h3>
<p>For this region we have a highly curated truth set for the mother available. It originates from the <a href="https://www.illumina.com/platinumgenomes.html">Illumina Platinum truth set</a>. You can find it at <code>data/variants/NA12878.vcf.gz</code></p>
<p>To check how well we did, we&rsquo;d first need to extract a vcf with only the information of the mother.</p>
<p><strong>Exercise:</strong> Generate a script called <code>C14_extract_mother_only.sh</code> to extract variants that have at least one alternative allele in the mother from <code>variants/trio.filtered.vcf</code>. Use <code>gatk SelectVariants</code> with the arguments:</p>
<ul>
<li><code>--sample-name mother</code></li>
<li><code>--exclude-non-variants</code></li>
<li><code>--remove-unused-alternates</code></li>
</ul>
<p>In addition to the required arguments.</p>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><span class="filename">C14_extract_mother_only.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project/results/variants

gatk<span class="w"> </span>SelectVariants<span class="w"> </span><span class="se">\</span>
--variant<span class="w"> </span>trio.filtered.vcf<span class="w"> </span><span class="se">\</span>
--sample-name<span class="w"> </span>mother<span class="w"> </span><span class="se">\</span>
--exclude-non-variants<span class="w"> </span><span class="se">\</span>
--remove-unused-alternates<span class="w"> </span><span class="se">\</span>
--output<span class="w"> </span>mother.trio.filtered.vcf
</code></pre></div>
</details>
<p><strong>Exercise:</strong></p>
<p>A. How many variants are in the extracted vcf? How many of those are filtered out?</p>
<p>B. Compare our vcf with the curated truth set with the command below from a script called <code>C15_evaluate_concordance.sh</code>. How many SNPs didn&rsquo;t we detect?</p>
<div class="highlight"><pre><span></span><code>gatk<span class="w"> </span>Concordance<span class="w"> </span><span class="se">\</span>
--evaluation<span class="w"> </span>variants/mother.trio.filtered.vcf<span class="w"> </span><span class="se">\</span>
--truth<span class="w"> </span>data/variants/NA12878.vcf.gz<span class="w"> </span><span class="se">\</span>
--intervals<span class="w"> </span>chr20:10018000-10220000<span class="w"> </span><span class="se">\</span>
--summary<span class="w"> </span>variants/concordance.mother.trio.filtered
</code></pre></div>
<details class="done">
<summary>Answer</summary>
<p>To get the number of records per FILTER, we run:</p>
<div class="highlight"><pre><span></span><code>grep<span class="w"> </span>-v<span class="w"> </span><span class="s2">&quot;^#&quot;</span><span class="w"> </span>variants/mother.trio.filtered.vcf<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">7</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span>-c
</code></pre></div>
<p>gives:</p>
<div class="highlight"><pre><span></span><code>407 PASS
2 SOR3
</code></pre></div>
<p>So two records were filtered out, based on the Symmetric Odds Ratio (issues with strand bias).</p>
<p>Your script to evaluate the concordance:</p>
<div class="highlight"><span class="filename">C15_evaluate_concordance.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project

gatk<span class="w"> </span>Concordance<span class="w"> </span><span class="se">\</span>
--evaluation<span class="w"> </span>results/variants/mother.trio.filtered.vcf<span class="w"> </span><span class="se">\</span>
--truth<span class="w"> </span>data/variants/NA12878.vcf.gz<span class="w"> </span><span class="se">\</span>
--intervals<span class="w"> </span>chr20:10018000-10220000<span class="w"> </span><span class="se">\</span>
--summary<span class="w"> </span>results/variants/concordance.mother.trio.filtered
</code></pre></div>
<p>Check out the output with <code>cat</code>:</p>
<div class="highlight"><pre><span></span><code>cat<span class="w"> </span>variants/concordance.mother.trio.filtered
</code></pre></div>
<p>gives:</p>
<div class="highlight"><pre><span></span><code>type    TP      FP      FN      RECALL  PRECISION
SNP     319     5       9       0.973   0.985
INDEL   63      20      6       0.913   0.759
</code></pre></div>
<p>Showing that there were 9 false negatives, i.e. SNPs we didn&rsquo;t detect.</p>
</details>
<div class="admonition tip">
<p class="admonition-title">Recall &amp; precision</p>
<p>More info on the definition of recall and precision on this <a href="https://en.wikipedia.org/wiki/Precision_and_recall">wikipedia page</a></p>
</div>
<p><strong>Exercise:</strong> Check out the concordance of the mother with the truth set before filtering. Do this by generating two scripts:</p>
<ul>
<li><code>C16_extract_mother_before_filtering.sh</code>: to run <code>gatk SelectVariants</code> in order to get only variants from the mother from the unfiltered <code>trio.vcf</code>. </li>
<li><code>C17_evaluate_concordance_before_filtering.sh</code>: to run <code>gatk Concordance</code> on the selected variants. </li>
</ul>
<p>Did filtering improve the recall or precision?</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>We did the filtering on <code>trio.vcf</code>, therefore, you first have to extract the records that only apply to the mother by using <code>gatk SelectVariants</code>.</p>
<p>Also note that <code>trio.vcf</code> contains records other than SNPs and INDELs. Use <code>--select-type-to-include</code> to select only SNPs and INDELs.</p>
</div>
<details class="done">
<summary>Answer</summary>
<p>First select only SNPs and INDELs from the mother from the unfiltered vcf:</p>
<div class="highlight"><span class="filename">C16_extract_mother_before_filtering.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project/results/variants

gatk<span class="w"> </span>SelectVariants<span class="w"> </span><span class="se">\</span>
--variant<span class="w"> </span>trio.vcf<span class="w"> </span><span class="se">\</span>
--sample-name<span class="w"> </span>mother<span class="w"> </span><span class="se">\</span>
--exclude-non-variants<span class="w"> </span><span class="se">\</span>
--remove-unused-alternates<span class="w"> </span><span class="se">\</span>
--select-type-to-include<span class="w"> </span>INDEL<span class="w"> </span><span class="se">\</span>
--select-type-to-include<span class="w"> </span>SNP<span class="w"> </span><span class="se">\</span>
--output<span class="w"> </span>mother.trio.vcf
</code></pre></div>
<p>Get the concordance with the truth set:</p>
<div class="highlight"><span class="filename">C17_evaluate_concordance_before_filtering.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project

gatk<span class="w"> </span>Concordance<span class="w"> </span><span class="se">\</span>
--evaluation<span class="w"> </span>results/variants/mother.trio.vcf<span class="w"> </span><span class="se">\</span>
--truth<span class="w"> </span>data/variants/NA12878.vcf.gz<span class="w"> </span><span class="se">\</span>
--intervals<span class="w"> </span>chr20:10018000-10220000<span class="w"> </span><span class="se">\</span>
--summary<span class="w"> </span>results/variants/concordance.mother.trio
</code></pre></div>
<p>Which gives:</p>
<div class="highlight"><pre><span></span><code>type    TP      FP      FN      RECALL  PRECISION
SNP     319     7       9       0.973   0.979
INDEL   63      20      6       0.913   0.759
</code></pre></div>
<p>The precision for SNPs is slightly lower. Due to filtering, we removed two false positives.</p>
</details>









  




                

  <!-- Giscus -->
  <h2>Questions & Answers</h2>

    <script src="https://giscus.app/client.js"
            data-repo="sib-swiss/NGS-variants-training"
            data-repo-id="MDEwOlJlcG9zaXRvcnkzMTEzMTc0MDI="
            data-category="Q&A"
            data-category-id="DIC_kwDOEo5Tms4CQ1hn"
            data-mapping="pathname"
            data-strict="0"
            data-reactions-enabled="1"
            data-emit-metadata="0"
            data-input-position="top"
            data-theme="preferred_color_scheme"
            data-lang="en"
            crossorigin="anonymous"
            async>
    </script>
  <!-- Synchronize Giscus theme with palette -->
  <script>
    var giscus = document.querySelector("script[src*=giscus]")

    /* Set palette on initial load */
    var palette = __md_get("__palette")
    if (palette && typeof palette.color === "object") {
      var theme = palette.color.scheme === "slate" ? "dark" : "light"
      giscus.setAttribute("data-theme", theme) 


    }

    /* Register event handlers after documented loaded */
    document.addEventListener("DOMContentLoaded", function() {
      var ref = document.querySelector("[data-md-component=palette]")
      ref.addEventListener("change", function() {
        var palette = __md_get("__palette")
        if (palette && typeof palette.color === "object") {
          var theme = palette.color.scheme === "slate" ? "dark" : "light"

          /* Instruct Giscus to change theme */
          var frame = document.querySelector(".giscus-frame")
          frame.contentWindow.postMessage(
            { giscus: { setConfig: { theme } } },
            "https://giscus.app"
          )
        }
      })
    })
  </script>

              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2021-2022 SIB Swiss institute of Bioinformatics –

    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["announce.dismiss"], "search": "../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"default": "latest", "provider": "mike"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.af256bd8.min.js"></script>
      
    
  </body>
</html>