
<!doctype html>
<html lang="en" class="no-js">
  <head>
    

      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://sib-swiss.github.io/NGS-variants-training/2024.2/day1/alignment_advanced/">
      
      
        <link rel="prev" href="../alignment/">
      
      
        <link rel="next" href="../../day2/variant_calling_preparation/">
      
      
      <link rel="icon" href="../../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.33">
    
<script type="application/ld+json">
  {
    "@context": "https://schema.org/",
    "@type": "LearningResource",
    "@id": "https://sib-swiss.github.io/NGS-variants-training",
    "http://purl.org/dc/terms/conformsTo": {
        "@type": "CreativeWork",
        "@id": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE"
    },
    "description": "Course on detecting variants from next generation sequencing data",
    "keywords": "NGS, variant, SNP, INDEL, sequencing, SNV, gatk",
    "name": "NGS - variant analysis",
    "about": [
        {
            "@type": "DefinedTerm",
            "@id": "http://edamontology.org/topic_3168",
            "inDefinedTermSet": "http://edamontology.org",
            "termCode": "topic_3168",
            "name": "Sequencing"
        },
        {
            "@type": "DefinedTerm",
            "@id": "http://edamontology.org/topic_3227",
            "inDefinedTermSet": "http://edamontology.org",
            "termCode": "topic_3227",
            "name": "Variant calling"
        }
    ],
    "abstract": "The detection of genetic variation is of major interest in various disciplines spanning from ecology and evolution research to inherited disease discovery and precision oncology. Next generation sequencing (NGS) methods are very powerful for the detection of genomic variants. Thanks to its throughput and cost-efficiency it enables the detection of a large number of variants in a large number of samples. In this two-day course we will cover the steps from read alignment to variant calling and annotation. We will mainly focus on the detection of germline mutations by following the GATK best practices. ",
    "audience": "Biologists working with NGS data",
    "author": [
        {
            "@type": "Person",
            "name": "Geert van Geest",
            "email": "geert.vangeest@sib.swiss"
        },
        {
            "@type": "Person",
            "name": "Patricia Palagi"
        },
        {
            "@type": "Organization",
            "name": "SIB Swiss Institute of Bioinformatics"
        }
    ],
    "identifier": "https://doi.org/10.5281/zenodo.6457818",
    "license": "https://creativecommons.org/licenses/by/4.0/",
    "teaches": [
        "Understand important aspects of NGS and read alignment for variant analysis",
        "Perform a read alignment ready for variant analysis",
        "Perform variant calling according to GATK best practices",
        "Perform a variant annotation"
    ]
}
</script>

    
      
        <title>Read alignment - advanced - NGS - variant analysis</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.3cba04c6.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
<!-- Matomo -->
<script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
      var u="https://matomo.sib.swiss/";
      _paq.push(['setTrackerUrl', u+'matomo.php']);
      _paq.push(['setSiteId', '220']);
      var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
      g.async=true; g.src=u+'matomo.js'; s.parentNode.insertBefore(g,s);
    })();
  </script>
  <!-- End Matomo Code -->
  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#learning-outcomes" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
        <aside class="md-banner">
          <div class="md-banner__inner md-grid md-typeset">
            
              <button class="md-banner__button md-icon" aria-label="Don't show this again">
                
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
              </button>
            
            
  <b>Issues, questions or feedback? &#128591; &#128640; </b> Use the "Questions and Answers" section at the bottom &#128071; of every page

          </div>
          
            <script>var content,el=document.querySelector("[data-md-component=announce]");el&&(content=el.querySelector(".md-typeset"),__md_hash(content.innerHTML)===__md_get("__announce")&&(el.hidden=!0))</script>
          
        </aside>
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
          <aside class="md-banner md-banner--warning">
            <div class="md-banner__inner md-grid md-typeset">
              
  You're not viewing the latest version.
  <a href="../../.."> 
    <strong>Click here to go to latest.</strong>
  </a>

            </div>
            <script>var el=document.querySelector("[data-md-component=outdated]"),outdated=__md_get("__outdated",sessionStorage);!0===outdated&&el&&(el.hidden=!1)</script>
          </aside>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="NGS - variant analysis" class="md-header__button md-logo" aria-label="NGS - variant analysis" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            NGS - variant analysis
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Read alignment - advanced
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/sib-swiss/NGS-variants-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/NGS-variants-training
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="NGS - variant analysis" class="md-nav__button md-logo" aria-label="NGS - variant analysis" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    NGS - variant analysis
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/sib-swiss/NGS-variants-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/NGS-variants-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../precourse/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Precourse preparations
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../course_schedule/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Course schedule
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Day 1
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Day 1
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../server_login/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Setup
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reproducibility/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Reproducibility
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../alignment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Read alignment - basics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Read alignment - advanced
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Read alignment - advanced
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    <span class="md-ellipsis">
      Learning outcomes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    <span class="md-ellipsis">
      Material
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    <span class="md-ellipsis">
      Exercises
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-adding-readgroups" class="md-nav__link">
    <span class="md-ellipsis">
      1. Adding readgroups
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-mark-duplicates" class="md-nav__link">
    <span class="md-ellipsis">
      2. Mark duplicates
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-indexing" class="md-nav__link">
    <span class="md-ellipsis">
      3. Indexing
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-recap-mother-only" class="md-nav__link">
    <span class="md-ellipsis">
      4. Recap - mother only
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-apply-it-on-all-three-samples-with-pipes-and-loops" class="md-nav__link">
    <span class="md-ellipsis">
      5. Apply it on all three samples with pipes and loops
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-recap-all-samples" class="md-nav__link">
    <span class="md-ellipsis">
      6. Recap - all samples
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Day 2
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Day 2
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/variant_calling_preparation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant calling - preparation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/variant_calling/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant calling
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/visualisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Visualisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/filtering_evaluation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Filtering & evaluation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Annotation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    <span class="md-ellipsis">
      Learning outcomes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    <span class="md-ellipsis">
      Material
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    <span class="md-ellipsis">
      Exercises
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-adding-readgroups" class="md-nav__link">
    <span class="md-ellipsis">
      1. Adding readgroups
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-mark-duplicates" class="md-nav__link">
    <span class="md-ellipsis">
      2. Mark duplicates
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-indexing" class="md-nav__link">
    <span class="md-ellipsis">
      3. Indexing
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-recap-mother-only" class="md-nav__link">
    <span class="md-ellipsis">
      4. Recap - mother only
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-apply-it-on-all-three-samples-with-pipes-and-loops" class="md-nav__link">
    <span class="md-ellipsis">
      5. Apply it on all three samples with pipes and loops
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-recap-all-samples" class="md-nav__link">
    <span class="md-ellipsis">
      6. Recap - all samples
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
  
                  

  
  


  <h1>Read alignment - advanced</h1>

<h2 id="learning-outcomes">Learning outcomes</h2>
<p><strong>After having completed this chapter you will be able to:</strong></p>
<ul>
<li>Use <code>samtools</code> to mark duplicates from an alignment file</li>
<li>Use <code>samtools</code> to add readgroups to an alignment file</li>
<li>Use a for loop in bash to perform the same operation on a range of files</li>
<li>Use <code>samtools</code> in a pipe to efficiently do multiple operations on an alignment file in a single command</li>
</ul>
<h2 id="material">Material</h2>
<ul>
<li><code>samtools</code> <a href="http://www.htslib.org/doc/samtools.html">documentation</a></li>
</ul>
<h2 id="exercises">Exercises</h2>
<h3 id="1-adding-readgroups">1. Adding readgroups</h3>
<p>During several steps of variant calling <code>gatk</code> uses read group information. For each read, this gives information on the sequencing platform, the library, the lane and of course the sample. Have a look at the description of the different levels of read group information <code>gatk</code> uses <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890671-Read-groups">here</a>. </p>
<p><strong>Exercise:</strong> The documentation mentions several read group fields that are used by <code>gatk</code>. Have a look at the <code>fastq</code> header. Does that give you the information that is required? Do we have that information for our sample? Can you specify it for our sample?</p>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>You can have a look at the first few entries in the <code>fastq</code> file with:</p>
<div class="highlight"><pre><span></span><code>zcat<span class="w"> </span>mother_R1.fastq.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>head
</code></pre></div>
</div>
<details class="done">
<summary>Answer</summary>
<p>Most of the information you should now based on the experimental design, the rest you can find in the <a href="https://en.wikipedia.org/wiki/FASTQ_format">fastq header</a>:</p>
<ul>
<li><code>PL</code>: the platform. Should be quite obvious; you usually you have this information. For us, this would be <code>ILLUMINA</code></li>
<li><code>SM</code>: the sample. All alignments that have reads coming from the same individual should have the same identifier in this field. For us, this would be <code>mother</code>. </li>
<li><code>LB</code>: library identifier. Molecular duplicates only exist within a library. If a single library was sequenced on multiple lanes, it is important to track this information. In our case, we have sequenced only one library, so you can specify it with e.g. <code>lib1</code>. </li>
<li><code>PU</code>: platform unit. This field is used to identify the sequencing lane. The documentation tells us we should specify it as <code>[FLOWCELL].[LANE].[SAMPLE BARCODE]</code>. The header of the first entry in our fastq file looks like this: <code>@H0164ALXX140820:2:1101:2136:40460/1</code>. Where the flowcell ID is <code>H0164</code> and the lane <code>2</code>. This formatting is specific to Broad Genomic Services pipelines, and not very common nowadays. Here the sample barcode is added to the flowcell ID, and is therefore specified as ALXX140820. We can therefore specify it with <code>H0164.2.ALXX140820</code>. </li>
<li><code>ID</code>: read group id. If you don&rsquo;t have specific information on the flowcell and lane (specified with <code>PU</code>), you can use this field to specify a unique unit that is used for e.g. base quality score recalibration. This often a combination of a flow cell identifier and a lane. In our case this could be <code>H0164.2</code></li>
</ul>
</details>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>More modern output of an Illumina sequencer looks e.g. like this (example on <a href="https://en.wikipedia.org/wiki/FASTQ_format">Wikipedia</a>):</p>
<div class="highlight"><pre><span></span><code>@EAS139:136:FC706VJ:2:2104:15343:197393 1:Y:18:ATCACG
</code></pre></div>
<p>Here, e.g. the <code>PU</code> field would be <code>FC706VJ.2.ATCACG</code></p>
</div>
<p><strong>Exercise:</strong> Have a look at the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037226472-AddOrReplaceReadGroups-Picard-">documentation</a> of <code>AddOrReplaceReadGroups</code>. Specify the required arguments, and run the command. Do this from a script called <code>B05_add_readgroups.sh</code> (in <code>~/project/scripts/B-mother_only</code>).</p>
<details class="done">
<summary>Answer</summary>
<p>We can use the answers of the previous exercise, and use them in the command:</p>
<div class="highlight"><span class="filename">B05_add_readgroups.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project/results

gatk<span class="w"> </span>AddOrReplaceReadGroups<span class="w"> </span><span class="se">\</span>
--INPUT<span class="w"> </span>alignments/mother.bam<span class="w"> </span><span class="se">\</span>
--OUTPUT<span class="w"> </span>alignments/mother.rg.bam<span class="w"> </span><span class="se">\</span>
--RGLB<span class="w"> </span>lib1<span class="w"> </span><span class="se">\</span>
--RGPU<span class="w"> </span>H0164.2.ALXX140820<span class="w"> </span><span class="se">\</span>
--RGPL<span class="w"> </span>ILLUMINA<span class="w"> </span><span class="se">\</span>
--RGSM<span class="w"> </span>mother<span class="w"> </span><span class="se">\</span>
--RGID<span class="w"> </span>H0164.2
</code></pre></div>
</details>
<p><strong>Exercise:</strong> Compare the header and first alignments of <code>mother.bam</code> and <code>mother.rg.bam</code>. Notice any differences?</p>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>You can view the header with</p>
<div class="highlight"><pre><span></span><code>samtools view -H &lt;alignment.bam&gt;
</code></pre></div>
<p>And the first few alignments with</p>
<div class="highlight"><pre><span></span><code>samtools view &lt;alignment.bam&gt; | head
</code></pre></div>
</div>
<details class="done">
<summary>Answer</summary>
<p>Compared to the header of <code>mother.markdup.bam</code>, the header of <code>mother.markdup.rg.bam</code> contains an extra line starting with <code>@RG</code>:</p>
<div class="highlight"><pre><span></span><code>@RG     ID:H0164.2      LB:lib1 PL:ILLUMINA     SM:mother       PU:H0164.2.ALXX140820
</code></pre></div>
<p>In the alignment records, a tag was added at the very end of each line: <code>RG:Z:H0164.2</code>. Note that all fields (<code>LB</code>, <code>PU</code>, etc.) are related to <code>ID</code>. So for each read only <code>ID</code> is specified and all other fields can be deducted from that. </p>
</details>
<h3 id="2-mark-duplicates">2. Mark duplicates</h3>
<p>Now that we have specified read groups, we can mark the duplicates with <code>gatk MarkDuplicates</code>. </p>
<p><strong>Exercise:</strong> Have a look at the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-">documentation</a>, and run <code>gatk MarkDuplicates</code> with the three required arguments. Do this from a script called <code>B06_mark_duplicates.sh</code> (in <code>~/project/scripts/B-mother_only</code>). </p>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><span class="filename">B06_mark_duplicates.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project/results

gatk<span class="w"> </span>MarkDuplicates<span class="w"> </span><span class="se">\</span>
--INPUT<span class="w"> </span>alignments/mother.rg.bam<span class="w"> </span><span class="se">\</span>
--OUTPUT<span class="w"> </span>alignments/mother.rg.md.bam<span class="w"> </span><span class="se">\</span>
--METRICS_FILE<span class="w"> </span>alignments/marked_dup_metrics_mother.txt<span class="w"> </span>
</code></pre></div>
</details>
<p><strong>Exercise:</strong> Run <code>samtools flagstat</code> on the alignment file with marked duplicates, and write the output to a file called <code>mother.rg.md.bam.flagstat</code>. Create a script called <code>B07_get_alignment_stats_after_md.sh</code> (in <code>~/project/scripts/B-mother_only</code>). How many reads were marked as duplicate?</p>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><span class="filename">B07_get_alignment_stats_after_md.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project/results/alignments

samtools<span class="w"> </span>flagstat<span class="w"> </span>mother.rg.md.bam<span class="w"> </span>&gt;<span class="w"> </span>mother.rg.md.bam.flagstat
</code></pre></div>
<p>Gives:</p>
<p><div class="highlight"><pre><span></span><code>133477 + 0 in total (QC-passed reads + QC-failed reads)
0 + 0 secondary
317 + 0 supplementary
17329 + 0 duplicates
132892 + 0 mapped (99.56% : N/A)
133160 + 0 paired in sequencing
66580 + 0 read1
66580 + 0 read2
131470 + 0 properly paired (98.73% : N/A)
131990 + 0 with itself and mate mapped

585 + 0 singletons (0.44% : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ&gt;=5)
</code></pre></div>
Which tells us that 17329 reads were marked as duplicate.</p>
</details>
<h3 id="3-indexing">3. Indexing</h3>
<p>To look up specific alignments, it is convenient to have your alignment file indexed. An indexing can be compared to a kind of &lsquo;phonebook&rsquo; of your sequence alignment file. Indexing can be done with <code>samtools</code> as well, but it first needs to be sorted on coordinate (i.e. the alignment location). You can do it like this:</p>
<div class="highlight"><pre><span></span><code>samtools<span class="w"> </span>index<span class="w"> </span>&lt;bam<span class="w"> </span>file&gt;
</code></pre></div>
<p><strong>Exercise</strong>: Create a script called <code>B08_index_alignment.sh</code> (in <code>~/project/scripts/B-mother_only</code>) to perform the alignment.</p>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><span class="filename">B08_index_alignment.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project/results/alignments/

samtools<span class="w"> </span>index<span class="w"> </span>mother.rg.md.bam
</code></pre></div>
</details>
<h3 id="4-recap-mother-only">4. Recap - mother only</h3>
<p>Now we have performed now the following steps on the sample <code>mother</code>:</p>
<ul>
<li>Read alignment</li>
<li>Adding readgroups</li>
<li>Marking of duplicates</li>
<li>Indexing</li>
</ul>
<p>Your <code>scripts</code> directory should look like this:</p>
<div class="highlight"><pre><span></span><code>├── A-prepare_references
│   ├── A01_download_course_data.sh
│   └── A02_create_bwa_index.sh
├── B-mother_only
│   ├── B01_alignment.sh
│   ├── B02_get_alignment_statistics.sh
│   ├── B03_sort_alignment.sh
│   ├── B04_compress_alignment.sh
│   ├── B05_add_readgroups.sh
│   ├── B06_mark_duplicates.sh
│   ├── B07_get_alignment_stats_after_md.sh
│   └── B08_index_alignment.sh
└── C-all_samples
</code></pre></div>
<h3 id="5-apply-it-on-all-three-samples-with-pipes-and-loops">5. Apply it on all three samples with pipes and loops</h3>
<p>We now apply these steps to all three samples. In order to do that, we combine the alignment with the sorting and compression in one command. We can do that with piping the output of <code>bwa</code> to <code>samtools sort</code> and <code>samtools view</code>, like this:</p>
<div class="highlight"><pre><span></span><code><span class="nv">SAMPLE</span><span class="o">=</span><span class="s2">&quot;mother&quot;</span>

bwa<span class="w"> </span>mem<span class="w"> </span>data/reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa<span class="w"> </span><span class="se">\</span>
data/fastq/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>_R1.fastq.gz<span class="w"> </span><span class="se">\</span>
data/fastq/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>_R2.fastq.gz<span class="w"> </span><span class="se">\</span>
<span class="p">|</span><span class="w"> </span>samtools<span class="w"> </span>sort<span class="w"> </span><span class="se">\</span>
<span class="p">|</span><span class="w"> </span>samtools<span class="w"> </span>view<span class="w"> </span>-bh<span class="w"> </span>&gt;<span class="w"> </span>results/alignments/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>.bam
</code></pre></div>
<p><strong>Exercise</strong>: Make a directory in the scripts directory <code>C-all_samples</code> (so <code>~/project/scripts/C-all_samples</code>). In here, create a script called <code>C01_alignment_sorting_compression.sh</code>. Within that script use the above snippet to make a loop that performs the alignment, sorting and compression for all three samples (i.e. <code>mother</code>, <code>father</code> and <code>son</code>).</p>
<details class="done">
<summary>Answer</summary>
<p>Your <code>scripts</code> directory should look like:</p>
<div class="highlight"><pre><span></span><code>scripts
├── A-prepare_references
│   ├── A01_download_course_data.sh
│   └── A02_create_bwa_index.sh
├── B-mother_only
│   ├── B01_alignment.sh
│   ├── B02_get_alignment_statistics.sh
│   ├── B03_sort_alignment.sh
│   ├── B04_compress_alignment.sh
│   ├── B05_add_readgroups.sh
│   ├── B06_mark_duplicates.sh
│   ├── B07_get_alignment_stats_after_md.sh
│   └── B08_index_alignment.sh
└── C-all_samples
    └── C01_alignment_sorting_compression.sh
</code></pre></div>
<p>And the script: </p>
<div class="highlight"><span class="filename">C01_alignment_sorting_compression.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project

<span class="k">for</span><span class="w"> </span>SAMPLE<span class="w"> </span><span class="k">in</span><span class="w"> </span>mother<span class="w"> </span>father<span class="w"> </span>son
<span class="k">do</span>
<span class="w">    </span>bwa<span class="w"> </span>mem<span class="w"> </span>data/reference/Homo_sapiens.GRCh38.dna.chromosome.20.fa<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>data/fastq/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>_R1.fastq.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>data/fastq/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>_R2.fastq.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>samtools<span class="w"> </span>sort<span class="w"> </span><span class="se">\</span>
<span class="w">    </span><span class="p">|</span><span class="w"> </span>samtools<span class="w"> </span>view<span class="w"> </span>-bh<span class="w"> </span>&gt;<span class="w"> </span>results/alignments/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>.bam
<span class="k">done</span>
</code></pre></div>
</details>
<p>Now we continue with adding the readgroups. For each sample, we have to add specific information to the different readgroup fields. We can do that by looping over a tab delimited file with sample-specific information in each row. Let&rsquo;s create that tab-delimited file. </p>
<p><strong>Exercise</strong> Generate a tab-delimited file called <code>sample_rg_fields.txt</code> and store it in <code>~/project/results/</code>. In this file, each line should represent a sample (mother, father and son), and you specify the <code>SM</code>, <code>LB</code>, <code>PU</code> and <code>ID</code> fields. E.g., the first line (for &lsquo;mother&rsquo;) would look like:</p>
<div class="highlight"><pre><span></span><code>mother  lib1    H0164.2.ALXX140820  H0164.2
</code></pre></div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Make sure to add a newline (<span class="keys"><kbd class="key-enter">Enter</kbd></span>) at the end of the file. Otherwise a loop will stop at the second-last line.</p>
</div>
<details class="done">
<summary>Answer</summary>
<p>Your file should look like this:</p>
<div class="highlight"><pre><span></span><code>mother  lib1    H0164.2.ALXX140820  H0164.2
father  lib2    H0164.3.ALXX140820  H0164.3
son lib3    H0164.6.ALXX140820  H0164.6
</code></pre></div>
</details>
<p><strong>Exercise</strong> Generate a script called <code>C02_add_readgroups.sh</code> (in <code>~/project/scripts/C-all_samples</code>) to loop over the tab-delimited file (have a look at the last exercise in <a href="../server_login#loops">Setup</a>), and add the correct readgroups to the bam file of each sample with <code>gatk AddOrReplaceReadGroups</code>. </p>
<details class="hint">
<summary>Hint</summary>
<p>Try to just print the variables from a loop in order to check to see whether the loop performs according to your expectation. E.g.:</p>
<div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>~/project/results

cat<span class="w"> </span>sample_rg_fields.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="k">while</span><span class="w"> </span><span class="nb">read</span><span class="w"> </span>SAMPLE<span class="w"> </span>LB<span class="w"> </span>PU<span class="w"> </span>ID
<span class="k">do</span>
<span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="nv">$SAMPLE</span><span class="w"> </span><span class="nv">$LB</span><span class="w"> </span><span class="nv">$PU</span><span class="w"> </span><span class="nv">$ID</span>
<span class="k">done</span>
</code></pre></div>
</details>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><span class="filename">C02_add_readgroups.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project/results

cat<span class="w"> </span>sample_rg_fields.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="k">while</span><span class="w"> </span><span class="nb">read</span><span class="w"> </span>SAMPLE<span class="w"> </span>LB<span class="w"> </span>PU<span class="w"> </span>ID
<span class="k">do</span>
<span class="w">    </span>gatk<span class="w"> </span>AddOrReplaceReadGroups<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--INPUT<span class="w"> </span>alignments/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>.bam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--OUTPUT<span class="w"> </span>alignments/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>.rg.bam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--RGLB<span class="w"> </span><span class="s2">&quot;</span><span class="nv">$LB</span><span class="s2">&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--RGPU<span class="w"> </span><span class="s2">&quot;</span><span class="nv">$PU</span><span class="s2">&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--RGPL<span class="w"> </span>ILLUMINA<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--RGSM<span class="w"> </span><span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--RGID<span class="w"> </span><span class="s2">&quot;</span><span class="nv">$ID</span><span class="s2">&quot;</span>
<span class="k">done</span><span class="w"> </span>
</code></pre></div>
</details>
<p>As final step, we will mark the duplicates and perform the indexing for the three samples. </p>
<p><strong>Exercise:</strong> Generate two scripts called <code>C03_mark_duplicates.sh</code> and <code>C04_index_alignments.sh</code>, in which you loop over the sample names and  perform the respective calculations. You can use <code>B06_mark_duplicates.sh</code> and <code>B08_index_alignment.sh</code> as a template. </p>
<details class="done">
<summary>Answer</summary>
<div class="highlight"><span class="filename">C03_mark_duplicates.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project/results

<span class="k">for</span><span class="w"> </span>SAMPLE<span class="w"> </span><span class="k">in</span><span class="w"> </span>mother<span class="w"> </span>father<span class="w"> </span>son
<span class="k">do</span>
<span class="w">    </span>gatk<span class="w"> </span>MarkDuplicates<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--INPUT<span class="w"> </span>alignments/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>.rg.bam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--OUTPUT<span class="w"> </span>alignments/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>.rg.md.bam<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--METRICS_FILE<span class="w"> </span>alignments/marked_dup_metrics_<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>.txt<span class="w"> </span>
<span class="k">done</span>
</code></pre></div>
<div class="highlight"><span class="filename">C04_index_alignment.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/project/results

<span class="k">for</span><span class="w"> </span>SAMPLE<span class="w"> </span><span class="k">in</span><span class="w"> </span>mother<span class="w"> </span>father<span class="w"> </span>son
<span class="k">do</span>
<span class="w">    </span>samtools<span class="w"> </span>index<span class="w"> </span>alignments/<span class="s2">&quot;</span><span class="nv">$SAMPLE</span><span class="s2">&quot;</span>.rg.md.bam
<span class="k">done</span><span class="w">  </span>
</code></pre></div>
</details>
<h3 id="6-recap-all-samples">6. Recap - all samples</h3>
<p>Now, we have performed on all three samples:</p>
<ul>
<li>Alignment</li>
<li>Sorting</li>
<li>Compression</li>
<li>Adding readgroups</li>
<li>Marking of duplicates</li>
<li>Indexing</li>
</ul>
<p>Your scripts directory should look like this:</p>
<div class="highlight"><pre><span></span><code>scripts
├── A-prepare_references
│   ├── A01_download_course_data.sh
│   └── A02_create_bwa_index.sh
├── B-mother_only
│   ├── B01_alignment.sh
│   ├── B02_get_alignment_statistics.sh
│   ├── B03_sort_alignment.sh
│   ├── B04_compress_alignment.sh
│   ├── B05_add_readgroups.sh
│   ├── B06_mark_duplicates.sh
│   ├── B07_get_alignment_stats_after_md.sh
│   └── B08_index_alignment.sh
└── C-all_samples
    ├── C01_alignment_sorting_compression.sh
    ├── C02_add_readgroups.sh
    ├── C03_mark_duplicates.sh
    └── C04_index_alignment.sh
</code></pre></div>









  




                

  <!-- Giscus -->
  <h2>Questions & Answers</h2>

    <script src="https://giscus.app/client.js"
            data-repo="sib-swiss/NGS-variants-training"
            data-repo-id="MDEwOlJlcG9zaXRvcnkzMTEzMTc0MDI="
            data-category="Q&A"
            data-category-id="DIC_kwDOEo5Tms4CQ1hn"
            data-mapping="pathname"
            data-strict="0"
            data-reactions-enabled="1"
            data-emit-metadata="0"
            data-input-position="top"
            data-theme="preferred_color_scheme"
            data-lang="en"
            crossorigin="anonymous"
            async>
    </script>
  <!-- Synchronize Giscus theme with palette -->
  <script>
    var giscus = document.querySelector("script[src*=giscus]")

    /* Set palette on initial load */
    var palette = __md_get("__palette")
    if (palette && typeof palette.color === "object") {
      var theme = palette.color.scheme === "slate" ? "dark" : "light"
      giscus.setAttribute("data-theme", theme) 


    }

    /* Register event handlers after documented loaded */
    document.addEventListener("DOMContentLoaded", function() {
      var ref = document.querySelector("[data-md-component=palette]")
      ref.addEventListener("change", function() {
        var palette = __md_get("__palette")
        if (palette && typeof palette.color === "object") {
          var theme = palette.color.scheme === "slate" ? "dark" : "light"

          /* Instruct Giscus to change theme */
          var frame = document.querySelector(".giscus-frame")
          frame.contentWindow.postMessage(
            { giscus: { setConfig: { theme } } },
            "https://giscus.app"
          )
        }
      })
    })
  </script>

              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2021-2022 SIB Swiss institute of Bioinformatics –

    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["announce.dismiss"], "search": "../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"default": "latest", "provider": "mike"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.af256bd8.min.js"></script>
      
    
  </body>
</html>